internal/operator/snapshotlease/snapshotlease.go

```diff
@@ -0,0 +1,201 @@
+	}
+	fullSnapshotObjectKey := client.ObjectKey{Name: etcd.GetFullSnapshotLeaseName(), Namespace: etcd.Namespace}
+	fullSnapshotLease, err := r.getLease(ctx, fullSnapshotObjectKey)
+	if err != nil {
+		return resourceNames, &druiderr.DruidError{
+			Code:      ErrGetSnapshotLease,
+			Cause:     err,
+			Operation: "GetExistingResourceNames",
+			Message:   fmt.Sprintf("Error getting full snapshot lease: %v for etcd: %v", fullSnapshotObjectKey, etcd.GetNamespaceName()),
+		}
+	}
+	if fullSnapshotLease != nil && metav1.IsControlledBy(fullSnapshotLease, etcd) {
+		resourceNames = append(resourceNames, fullSnapshotLease.Name)
+	}
+	return resourceNames, nil
+}
+
+// Sync creates or updates the snapshot leases for the given Etcd.
+// If backups are disabled for the Etcd resource, any existing snapshot leases are deleted.
+func (r _resource) Sync(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) error {
+	if !etcd.IsBackupStoreEnabled() {
+		ctx.Logger.Info("Backup has been disabled. Triggering deletion of snapshot leases")
+		return r.deleteAllSnapshotLeases(ctx, etcd, func(err error) error {
+			return druiderr.WrapError(err,
+				ErrSyncSnapshotLease,
+				"Sync",
+				fmt.Sprintf("Failed to delete existing snapshot leases due to backup being disabled for etcd: %v", etcd.GetNamespaceName()))
+		})
+	}
+
+	objectKeys := getObjectKeys(etcd)
+	syncTasks := make([]utils.OperatorTask, len(objectKeys))
+
+	for i, objKey := range objectKeys {
+		objKey := objKey // capture the range variable
+		syncTasks[i] = utils.OperatorTask{
+			Name: "CreateOrUpdate-" + objKey.String(),
+			Fn: func(ctx component.OperatorContext) error {
+				return r.doCreateOrUpdate(ctx, etcd, objKey)
+			},
+		}
+	}
+	return errors.Join(utils.RunConcurrently(ctx, syncTasks)...)
+}
+
+// TriggerDelete triggers the deletion of the snapshot leases for the given Etcd.
+func (r _resource) TriggerDelete(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) error {
+	ctx.Logger.Info("Triggering deletion of snapshot leases")
+	if err := r.deleteAllSnapshotLeases(ctx, etcd, func(err error) error {
+		return druiderr.WrapError(err,
+			ErrDeleteSnapshotLease,
+			"TriggerDelete",
+			fmt.Sprintf("Failed to delete snapshot leases for etcd: %v", etcd.GetNamespaceName()))
+	}); err != nil {
+		return err
+	}
+	ctx.Logger.Info("deleted", "component", "snapshot-leases")
+	return nil
+}
+
+func (r _resource) deleteAllSnapshotLeases(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd, wrapErrFn func(error) error) error {
+	if err := r.client.DeleteAllOf(ctx,
+		&coordinationv1.Lease{},
+		client.InNamespace(etcd.Namespace),
+		client.MatchingLabels(getSelectorLabelsForAllSnapshotLeases(etcd))); err != nil {
+		return wrapErrFn(err)
+	}
+	return nil
+}
+
+func (r _resource) getLease(ctx context.Context, objectKey client.ObjectKey) (*coordinationv1.Lease, error) {
+	lease := &coordinationv1.Lease{}
+	if err := r.client.Get(ctx, objectKey, lease); err != nil {
```

Comment:
This could be metadata-only get instead of the full object.

Reply:
done